# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: third_party/rust/sync15/src/engine/changeset.rs
# Commit: 57b2cda7e0db
# Full Hash: 57b2cda7e0db8b07fcc1e09113ce26add178c68a
# Author: Sammy Khamis <skhamis@mozilla.com>
# Date: 2023-03-01 03:56:13
# Regressor Bug: 1818349
# File Overlap Count: 23
# Overlapping Files: third_party/rust/sync15/src/client/coll_update.rs, third_party/rust/sync15/src/client/mod.rs, third_party/rust/sync15/.cargo-checksum.json, third_party/rust/sync15/src/client/sync.rs, Cargo.toml
# Description:
#   Bug 1818349 Part 1: Vendor new version of application-services r=markh
#   
#   Differential Revision: https://phabricator.services.mozilla.com/D171181
# ==============================================================================

diff -r eaf9f12a1481 -r 57b2cda7e0db third_party/rust/sync15/src/engine/changeset.rs
--- a/third_party/rust/sync15/src/engine/changeset.rs	Tue Feb 28 04:22:35 2023 +0000
+++ b/third_party/rust/sync15/src/engine/changeset.rs	Tue Feb 28 05:14:29 2023 +0000
@@ -3,40 +3,50 @@
  * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
 
 use crate::bso::{IncomingBso, OutgoingBso};
-use crate::ServerTimestamp;
+use crate::{CollectionName, ServerTimestamp};
 
-#[derive(Debug, Clone)]
-pub struct RecordChangeset<T> {
-    pub changes: Vec<T>,
-    /// For GETs, the last sync timestamp that should be persisted after
-    /// applying the records.
-    /// For POSTs, this is the XIUS timestamp.
+// Incoming and Outgoing changesets are almost identical except for the timestamp.
+// Separate types still helps avoid confusion with that timestamp, so they're split.
+#[derive(Debug)]
+pub struct IncomingChangeset {
+    pub changes: Vec<IncomingBso>,
+    /// The server timestamp of the collection.
     pub timestamp: ServerTimestamp,
-    pub collection: std::borrow::Cow<'static, str>,
+    pub collection: CollectionName,
 }
 
-pub type IncomingChangeset = RecordChangeset<IncomingBso>;
-pub type OutgoingChangeset = RecordChangeset<OutgoingBso>;
-
-impl<T> RecordChangeset<T> {
+impl IncomingChangeset {
     #[inline]
-    pub fn new(
-        collection: impl Into<std::borrow::Cow<'static, str>>,
-        timestamp: ServerTimestamp,
-    ) -> RecordChangeset<T> {
+    pub fn new(collection: CollectionName, timestamp: ServerTimestamp) -> Self {
         Self::new_with_changes(collection, timestamp, Vec::new())
     }
 
     #[inline]
     pub fn new_with_changes(
-        collection: impl Into<std::borrow::Cow<'static, str>>,
+        collection: CollectionName,
         timestamp: ServerTimestamp,
-        changes: Vec<T>,
-    ) -> RecordChangeset<T> {
-        RecordChangeset {
+        changes: Vec<IncomingBso>,
+    ) -> Self {
+        Self {
             changes,
             timestamp,
-            collection: collection.into(),
+            collection,
         }
     }
 }
+
+#[derive(Debug)]
+pub struct OutgoingChangeset {
+    pub changes: Vec<OutgoingBso>,
+    pub collection: CollectionName,
+}
+
+impl OutgoingChangeset {
+    #[inline]
+    pub fn new(collection: CollectionName, changes: Vec<OutgoingBso>) -> Self {
+        Self {
+            collection,
+            changes,
+        }
+    }
+}